R/infer_polars_dtype.R

```diff
@@ -0,0 +1,128 @@
+# TODO: link to the type mapping vignette
+#' Infer Polars DataType corresponding to a given R object
+#'
+#' This function is a helper function used to quickly find the [DataType] corresponding to an R object.
+#' In many cases, this function simply performs something like [`as_polars_series(x[0])`][as_polars_series].
+#' It is much faster than actually constructing a [Series] using the entire object.
+#' This function is similar to [nanoarrow::infer_nanoarrow_schema()].
+#'
+#' S3 objects based on atomic vectors or classes built on the vctrs package will work accurately
+#' if the S3 method of the [as_polars_series] function is defined.
+#' @inheritParams as_polars_series
+#' @param x An R object.
+#' @return A [polars DataType][DataType]
+#' @seealso
+#' - [as_polars_series()]
+#' @exmaples
+#' infer_polars_dtype(1:10)
+#' @export
+infer_polars_dtype <- function(x, ...) {
+  UseMethod("infer_polars_dtype")
+}
+
+infer_polars_dtype_default_impl <- function(x, ...) {
+  as_polars_series(x[0L]) |>
+    infer_polars_dtype(...)
+}
+
+#' @rdname infer_polars_dtype
+#' @export
+infer_polars_dtype.default <- function(x, ...) {
+  if (is.atomic(x)) {
+    infer_polars_dtype_default_impl(x, ...)
+  } else {
+    abort(
+      sprintf("Unsupported class for `infer_polars_dtype()`: %s", toString(class(x))),
+      call = parent.frame()
+    )
+  }
+}
+
+#' @rdname infer_polars_dtype
+#' @export
+infer_polars_dtype.polars_series <- function(x, ...) {
+  x$dtype
+}
+
+#' @rdname infer_polars_dtype
+#' @export
+infer_polars_dtype.polars_data_frame <- function(x, ...) {
+  pl$Struct(!!!x$collect_schema())
+}
+
+#' @rdname infer_polars_dtype
+#' @export
+infer_polars_dtype.polars_lazy_frame <- infer_polars_dtype.polars_data_frame
+
+#' @rdname infer_polars_dtype
+#' @export
+infer_polars_dtype.POSIXlt <- infer_polars_dtype_default_impl
+
+#' @rdname infer_polars_dtype
+#' @export
+infer_polars_dtype.NULL <- function(x, ...) {
+  if (missing(x)) {
+    abort("The `x` argument of `infer_polars_dtype()` can't be missing")
+  }
```

Comment:
Are there cases where we end up in this method if `x` is missing?

Reply:
I think I learned this pattern from `tibble:::as_tibble.NULL`.
Try running `tibble::as_tibble()`.

Reply:
I see. I think it would also make sense to put it before the call to `UseMethod()` but I'm fine with that